File: scripts/enemies/_enemy_handling.py
import random
import math

import pygame

from scripts.enemies import basic_enemy, basic_flying, bouncer
from scripts.enemies._enemy import costume_dimensions

light_gradient = None
enemy_animations = {}

enemies = [basic_enemy.BasicEnemy, basic_flying.BasicFlying, bouncer.Bouncer]
enemy_sizes = {}
enemy_costumes = {}

for enemy in enemies:
    enemy_sizes[enemy] = enemy.size_range
    enemy_costumes[enemy] = enemy.costume

eligible_enemies = {"white": [basic_enemy.BasicEnemy], "blue": [basic_enemy.BasicEnemy, basic_flying.BasicFlying], "red": [basic_enemy.BasicEnemy]}


def get_enemy(_terrain, player, nest_type, color, nest_health, nest_x, nest_y, nest_size):
    for i in range(20):
        angle = math.atan2(player.y - nest_y, player.x - nest_x) + (random.random() * 2 - 1) * math.pi/2
        r = random.randint(int(nest_size / 2 - 10), int(nest_size / 2 + 10))
        x, y = int(nest_x + r * math.cos(angle)), int(nest_y + r * math.sin(angle))

        eligible = eligible_enemies[nest_type]
        variant = eligible[random.randint(0, len(eligible) - 1)]

        size_min, size_max = enemy_sizes[variant]
        size = random.randint(size_min, size_max)
        width = size * costume_dimensions[enemy_costumes[variant]][0]
        height = size * costume_dimensions[enemy_costumes[variant]][1]

        new_enemy_rect = pygame.Rect(x - width / 2, y - height / 2, width, height)
        if not (_terrain.collide_rect(new_enemy_rect) or new_enemy_rect.colliderect(player.rect)):
            new_enemy = variant(_terrain.default_zooms, color, size, nest_health, x, y)
            new_enemy.spawn_particles(_terrain)
            return new_enemy
    return False

File: cpp_port/EnemyHandling.cpp
#include "EnemyHandling.h"
#include "BasicEnemy.h"
#include "BasicFlying.h"
#include "Bouncer.h"
#include "Util.h"
#include <cmath>
#include <vector>

namespace EnemyHandling {

namespace {
// was: costume_dimensions -- only "1" exists; reused here for the
// width/height pre-check before actually constructing an enemy (matches
// the Python computing width/height from enemy_costumes[variant] before
// constructing anything).
constexpr double kCostume1WidthFrac = 3.0 / 8.0;
constexpr double kCostume1HeightFrac = 3.0 / 4.0;

// was: eligible_enemies = {"white": [BasicEnemy], "blue": [BasicEnemy,
// BasicFlying], "red": [BasicEnemy]}
const std::vector<EnemyVariant>& eligibleEnemies(ChargeType nestType) {
    static const std::vector<EnemyVariant> white = { EnemyVariant::BasicEnemy_ };
    static const std::vector<EnemyVariant> blue = { EnemyVariant::BasicEnemy_, EnemyVariant::BasicFlying_ };
    static const std::vector<EnemyVariant> red = { EnemyVariant::BasicEnemy_ };
    switch (nestType) {
        case ChargeType::White: return white;
        case ChargeType::Blue: return blue;
        case ChargeType::Red: return red;
    }
    return white; // unreachable
}

int sizeMin(EnemyVariant v) {
    switch (v) {
        case EnemyVariant::BasicEnemy_: return BasicEnemy::kSizeMin;
        case EnemyVariant::BasicFlying_: return BasicFlying::kSizeMin;
        case EnemyVariant::Bouncer_: return Bouncer::kSizeMin;
    }
    return 0;
}
int sizeMax(EnemyVariant v) {
    switch (v) {
        case EnemyVariant::BasicEnemy_: return BasicEnemy::kSizeMax;
        case EnemyVariant::BasicFlying_: return BasicFlying::kSizeMax;
        case EnemyVariant::Bouncer_: return Bouncer::kSizeMax;
    }
    return 0;
}

std::unique_ptr<Enemy> construct(EnemyVariant v, SDL_Renderer* renderer, Color color,
                                  double size, double nestHealth, double x, double y) {
    switch (v) {
        case EnemyVariant::BasicEnemy_: return std::make_unique<BasicEnemy>(renderer, color, size, nestHealth, x, y);
        case EnemyVariant::BasicFlying_: return std::make_unique<BasicFlying>(renderer, color, size, nestHealth, x, y);
        case EnemyVariant::Bouncer_: return std::make_unique<Bouncer>(renderer, color, size, nestHealth, x, y);
    }
    return nullptr; // unreachable
}
} // namespace

// def get_enemy(_terrain, player, nest_type, color, nest_health, nest_x, nest_y, nest_size): ...
std::unique_ptr<Enemy> getEnemy(SDL_Renderer* renderer, Terrain& terrain,
                                 double playerX, double playerY, const Rect& playerRect,
                                 ChargeType nestType, Color color, double nestHealth,
                                 double nestX, double nestY, double nestSize) {
    for (int i = 0; i < 20; ++i) {
        double angle = std::atan2(playerY - nestY, playerX - nestX) + (Util::randomDouble() * 2 - 1) * M_PI / 2.0;
        int r = Util::randint(static_cast<int>(nestSize / 2.0 - 10), static_cast<int>(nestSize / 2.0 + 10));
        int x = static_cast<int>(nestX + r * std::cos(angle));
        int y = static_cast<int>(nestY + r * std::sin(angle));

        const auto& eligible = eligibleEnemies(nestType);
        EnemyVariant variant = eligible[Util::randint(0, static_cast<int>(eligible.size()) - 1)];

        double size = Util::randint(sizeMin(variant), sizeMax(variant));
        double width = size * kCostume1WidthFrac;
        double height = size * kCostume1HeightFrac;

        Rect newEnemyRect{
            static_cast<int>(x - width / 2.0), static_cast<int>(y - height / 2.0),
            static_cast<int>(width), static_cast<int>(height)
        };

        if (!(terrain.collideRect(newEnemyRect).hit || newEnemyRect.collideRect(playerRect))) {
            auto newEnemy = construct(variant, renderer, color, size, nestHealth, x, y);
            newEnemy->spawnParticles(terrain);
            return newEnemy;
        }
    }
    return nullptr;
}

} // namespace EnemyHandling
